demos/tests/correctness_checker.py

```diff
@@ -0,0 +1,135 @@
+
+from abc import ABC, abstractmethod
+from unittest import case
+
+class Demo(ABC):
+    def __init__(self, name, implementation):
+        self.name = name
+        self.implementation = implementation
+        self.subdirectory = name + '/' + implementation
+        self.results = {}
+    
+    @abstractmethod
+    def parser(self):
+        pass 
+
+    @abstractmethod
+    def checker(self):
+        pass
+
+class CPPDemo(Demo):
+    def __init__(self, name, implementation='cpp'):
+        super().__init__(name, implementation)
+        self.results = {}
+        pass
+
+    def parser(self):
+        fo = None
+        try:
+            fo = open('/tmp/' + self.subdirectory + '/results.log', 'r')
+        except IOError as err:
+            print("File error: " + str(err))
+        output = [i.rstrip() for i in fo.readlines()]
+        device = ''
+        case_index = ''
+        index = 1
+        while index < len(output):
+            if "Device" in output[index]:
+                device = output[index][output[index].find(":") + 1:]
+                if device not in self.results:
+                    self.results[device] = {}
+
+            if "CaseId" in output[index]:
+                case_index = output[index][output[index].find(":") + 1:]
+                if case_index not in self.results[device]:
+                    self.results[device][case_index] = {}
+            
+            if "Execution_time" in output[index]:
+                execution_time = output[index].split(':')[1]
+                if execution_time == '-1':
+                   while index < len(output) and 'Device' not in output[index]:
+                       index += 1 
+                   continue
+            
+            # Pase the raw data
+            while index < len(output) and 'ChannelId' in output[index]:
+                item = output[index][output[index].find('ChannelId'):].split(',')
+                # Channel ID
+                frame_results = {}
+                channel = item[0].split(':')
+                if channel[1] not in self.results[device][case_index]:
+                    self.results[device][case_index][channel[1]] = frame_results
+
+                # Frame ID
+                object_results = {}
+                frame = item[1].split(':')
+                if frame[1] not in self.results[device][case_index][channel[1]]:
+                    self.results[device][case_index][channel[1]][frame[1]] = object_results
+
+                # Object ID
+                label_prob_pos_results = []
+                objid = item[2].split(':')
+                if objid[1] not in self.results[device][case_index][channel[1]][frame[1]]:
+                    self.results[device][case_index][channel[1]][frame[1]][objid[1]] = label_prob_pos_results
+                self.results[device][case_index][channel[1]][frame[1]][objid[1]] = item[3:]
+                index += 1
+
+            index += 1
+
+    
+    def checker(self):
+        self.parser()
```

Comment:
It seems `parser()` will not be used somewhere outside the checker, so `checker()` and `parser()` should be combined

Reply:
Hi, I think the code lines of `checker()` will be too long if combining the parser function into `checker()`.